internal/pkg/agent/cmd/upgrade.go

```diff
@@ -53,20 +57,85 @@ func newUpgradeCommandWithArgs(_ []string, streams *cli.IOStreams) *cobra.Comman
 	cmd.Flags().String(flagPGPBytes, "", "PGP to use for package verification")
 	cmd.Flags().String(flagPGPBytesURI, "", "Path to a web location containing PGP to use for package verification")
 	cmd.Flags().String(flagPGPBytesPath, "", "Path to a file containing PGP to use for package verification")
+	cmd.Flags().BoolP(flagForce, "f", false, "Advanced option to force an upgrade on a fleet managed agent")
+	err := cmd.Flags().MarkHidden(flagForce)
+	if err != nil {
+		log.Fatal(err)
+	}
 
 	return cmd
 }
 
+type upgradeInput struct {
+	streams   *cli.IOStreams
+	cmd       *cobra.Command
+	args      []string
+	c         client.Client
+	agentInfo info.Agent
+	cFunc     confirmFunc
+}
+
+type confirmFunc func(string, bool) (bool, error)
+
 func upgradeCmd(streams *cli.IOStreams, cmd *cobra.Command, args []string) error {
 	c := client.New()
-	return upgradeCmdWithClient(streams, cmd, args, c)
+	agentInfo, err := info.NewAgentInfoWithLog(cmd.Context(), "error", false)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve agent info while tring to upgrade the agent: %w", err)
+	}
+	input := &upgradeInput{
+		streams,
+		cmd,
+		args,
+		c,
+		agentInfo,
+		cli.Confirm,
+	}
+	return upgradeCmdWithClient(input)
+}
+
+func shouldUpgrade(cmd *cobra.Command, agentInfo info.Agent, cFunc confirmFunc) (bool, error) {
+	if agentInfo.IsStandalone() {
+		return true, nil
+	}
+
+	if agentInfo.Unprivileged() {
+		return false, fmt.Errorf("upgrade command needs to be executed as root for fleet managed agents")
+	}
+
+	force, err := cmd.Flags().GetBool(flagForce)
+	if err != nil {
+		return false, fmt.Errorf("failed to retrieve command flag information while trying to upgrade the agent: %w", err)
+	}
+
+	if !force {
+		return false, fmt.Errorf("upgrading fleet managed agents is not supported")
+	}
+
+	cf, err := cFunc("Upgrading fleet managed agents is not supported. Would you still like to proceed?", false)
```

Comment:
@cmacknz Are we sure we want to introduce this confirmation (if I understand correctly, you proposed it [here](https://github.com/elastic/elastic-agent/issues/4890#issuecomment-2414789873))? This will make the command unusable in unattended scenarios, like Ansible mentioned in https://github.com/elastic/elastic-agent/issues/4890#issuecomment-2337891708.

It seems to me that passing the `--force` flag means that the user has already opted in and we should expect at this point that the user knows what they're doing. The additional confirmation sounds like we're asking the user the same thing twice: "Are you sure? But are you sure you're sure?".

On a separate but related note, I'm not fond of the `--force` flag being hidden from docs. I would rather have it not hidden, and have the docs clearly describe that using it is not recommended, what can go wrong and when it can be useful.

Reply:
+1 to getting rid of the confirmation for force.

The intent for Fleet managed agents, which already have an automated bulk upgrade mechanism, is for this to be a "use in case of emergency" option where the Fleet upgrades are broken. 

The official docs already state that this is not recommended for Fleet managed agents. https://www.elastic.co/guide/en/fleet/current/elastic-agent-cmd-options.html#elastic-agent-upgrade-command

Today using `elastic-agent upgrade` with a Fleet managed agent does not work at all because of a separate bug, so nobody is even able to use this.

The thing this is protecting us against is a situation where an agent has gone offline in Fleet because of a bug, and the way out is to upgrade the agent (or there is a bug in the Fleet upgrade process itself). Upgrading the agent should not be impossible in this circumstance.

The intent is to discourage use of `elastic-agent upgrade` outside of these circumstances, but if somebody is root, and they can discover this `--force` option exists, and they really want to do it, they can. That is fine.